tests/basic/clocks.rs

```diff
@@ -0,0 +1,415 @@
+        .collect::<Vec<_>>();
+
+    barrier.wait();
+    // Since all threads reached the barrier, everyone's clock must be nonzero, except for the dummy
+    check_clock(|i, c| (c > 0) == (i != 1));
+}
+
+#[test]
+fn clock_barrier_dfs() {
+    check_dfs(|| clock_barrier(4), None);
+}
+
+#[test]
+fn clock_barrier_pct() {
+    check_pct(|| clock_barrier(50), 1000, 3);
+}
+
+// Condvars
+#[test]
+fn clock_condvar_single() {
+    check_dfs(
+        || {
+            let lock = Arc::new(Mutex::new(false));
+            let cond = Arc::new(Condvar::new());
+
+            {
+                let lock = Arc::clone(&lock);
+                let cond = Arc::clone(&cond);
+                thread::spawn(move || {
+                    assert_eq!(me(), 1);
+                    *lock.lock().unwrap() = true;
+                    cond.notify_one();
+                });
+            }
+
+            let mut guard = lock.lock().unwrap();
+            while !*guard {
+                check_clock(|i, c| (c > 0) == (i == 0)); // spawned thread has not executed notify_one
+                guard = cond.wait(guard).unwrap();
+            }
+            check_clock(|i, c| (c > 0) == (i == 0 || i == 1));
+        },
+        None,
+    )
+}
+
+fn clock_condvar_notify_one(num_notifiers: usize, num_waiters: usize) {
+    let lock = Arc::new(Mutex::new(0usize));
+    let cond = Arc::new(Condvar::new());
+
+    for _ in 0..num_notifiers {
+        let lock = Arc::clone(&lock);
+        let cond = Arc::clone(&cond);
+        thread::spawn(move || {
+            assert!(me() <= num_notifiers);
+            *lock.lock().unwrap() = me();
+            cond.notify_one();
+        });
+    }
+
+    for _ in 0..num_waiters {
+        let lock = Arc::clone(&lock);
+        let cond = Arc::clone(&cond);
+        thread::spawn(move || {
+            let mut guard = lock.lock().unwrap();
+            while *guard == 0 {
+                check_clock(|i, c| !(i >= 1 && i <= num_notifiers) || (c == 0)); // no notifier has gone yet
+                guard = cond.wait(guard).unwrap();
+            }
+            // Note that since all the threads touch the lock, any of them may have preceded this thread.
+            // But we know for sure that the thread that unblocked us should causally precede us.
+            check_clock(|i, c| (i != *guard) || (c > 0));
+        });
+    }
+}
+
+#[test]
+fn clock_condvar_notify_one_dfs() {
+    check_dfs(|| clock_condvar_notify_one(1, 1), None);
+}
+
+#[test]
+fn clock_condvar_notify_one_pct() {
+    check_pct(|| clock_condvar_notify_one(10, 10), 10_000, 3);
+}
+
+fn clock_condvar_notify_all(num_waiters: usize) {
+    let lock = Arc::new(Mutex::new(0usize));
+    let cond = Arc::new(Condvar::new());
+
+    {
+        let lock = Arc::clone(&lock);
+        let cond = Arc::clone(&cond);
+        thread::spawn(move || {
+            assert_eq!(me(), 1);
+            *lock.lock().unwrap() = me();
+            cond.notify_all();
+        });
+    }
```

Comment:
Could move this after the waiters spawn and avoid spawning a new thread for it, which might make the DFS version of this test faster. But I guess we'd have to find a different way around the `me() == 1` thing.

Reply:
Thing is that I want the assertion on line 264 that checks that the waiter does not have causal dependency on the notifier as long as the lock reads zero.  This is easy to do with the extra thread (with taskid 1), since we can just check that the vector clock for thread-1 must be zero.  If we use thread 0 to do the notify, we can't use the same check, because thread 0 spawns all the waiters, so their clocks inherit its clock.  I could save a thread spawn by making the main thread be one of the waiters, but that would result in some code duplication, and i thought we could just live with a smaller dfs test.